crates/red_knot_python_semantic/src/types.rs

```diff
@@ -347,6 +368,36 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Given the type of an object that is iterated over in some way,
+    /// return the type of objects that are yielded by that iteration.
+    ///
+    /// E.g., for the following loop, given the type of `x`, infer the type of `y`:
+    /// ```python
+    /// for y in x:
+    ///     pass
+    /// ```
+    ///
+    /// Returns `None` if `self` represents a type that is not iterable.
+    fn iterate(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        // `self` represents the type of the iterable;
+        // `__iter__` and `__next__` are both looked up on the class of the iterable:
+        let type_of_class = self.to_type_of_class(db);
+        let dunder_iter_method = type_of_class.member(db, "__iter__");
+        if !dunder_iter_method.is_unbound() {
+            let iterator_ty = dunder_iter_method.call(db)?;
+            let dunder_next_method = iterator_ty.to_type_of_class(db).member(db, "__next__");
+            return dunder_next_method.call(db);
+        }
+        // Although it's not considered great practice,
+        // classes that define `__getitem__` are also iterable,
+        // even if they do not define `__iter__`.
+        //
+        // TODO this is only valid if the `__getitem__` method is annotated as
+        // accepting `int` or `SupportsIndex`
+        let dunder_get_item_method = type_of_class.member(db, "__getitem__");
+        dunder_get_item_method.call(db)
```

Comment:
Nit. Maybe add some blank lines to make it clearer that `__getitem__` is a separate lookup logic. Or use `if...else`
```suggestion
        if dunder_iter_method.is_bound() {
            let iterator_ty = dunder_iter_method.call(db)?;
            let dunder_next_method = iterator_ty.to_type_of_class(db).member(db, "__next__");
            dunder_next_method.call(db);
        } else {
	        // Although it's not considered great practice,
	        // classes that define `__getitem__` are also iterable,
	        // even if they do not define `__iter__`.
	        //
	        // TODO this is only valid if the `__getitem__` method is annotated as
	        // accepting `int` or `SupportsIndex`
	        let dunder_get_item_method = type_of_class.member(db, "__getitem__");
	        dunder_get_item_method.call(db)
        }
```

Reply:
Would this logic change once we have proper `Protocol` support?

Reply:
> Would this logic change once we have proper `Protocol` support?

I don't think so. It will become more useful when we have proper `Protocol` support, because most `__iter__` methods in Python are annotated as returning `Iterator[...]`, and `Iterator` is a protocol that desugars to "an instance of some class that has `__iter__` and `__next__` methods". So currently this logic does not recognise most Python iterators as actually being Python iterators, but it will do once we add `Protocol` support.

Once we have `Protocol` support, it's true that we _could_ "simplify" this logic by simply saying "anything that's a subtype of `Iterable[object]` counts as an iterable", and use that to determine whether something is iterable or not. I prefer my current logic, though, because:
- the `Iterator` protocol states that an iterator has to have `__next__` and `__iter__`, but this isn't actually required by the language semantics. It's considered good coding style for all iterators to also be iterable (have an `__iter__` method) but the language only requires them to have a `__next__` method.
- Only considering subtypes of `Iterable[object]` to be iterable ignores the fact that classes that use the old-style iteration protocol (using `__getitem__`) are also considered iterable by the language

Reply:
> Or use `if...else`

I tried that first, but Clippy insisted that I put the `__getitem__` logic before the `__iter__` logic:

```rs
if dunder_iter_method.is_unbound() {
    <__getitem__ logic>
} else {
    <__iter__ logic>
}
```

That seems very wrong to me, because the `__getitem__` logic is a fallback if the object doesn't have an `__iter__` method; the `__iter__` logic should definitely come first. We don't have an `is_bound()` method (only an `is_unbound()` method), and I'm not sure we should add one because we don't have a `Type::Bound` variant, only a `Type::Unbound` variant.

I'll add some more blank lines to make it clearer.